File: leetcode/62.unique-paths.py
# @leet start
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        # Use memoization. This is trivial using combinatorics though (I
        # think?).
        memo: dict[tuple[int, int], int] = {(m - 1, n - 1): 1}

        def num_unique_paths(row: int, col: int) -> int:
            # Lookup previous result
            if (row, col) in memo:
                return memo[(row, col)]

            # Compute result. For num ways we
            num_ways = 0

            if row < m - 1:
                num_ways += num_unique_paths(row + 1, col)

            if col < n - 1:
                num_ways += num_unique_paths(row, col + 1)

            memo[(row, col)] = num_ways

            return num_ways

        # Return the number of unique paths
        return num_unique_paths(0, 0)


# @leet end

File: leetcode/62.unique-paths.ts
// @leet start
function uniquePaths(m: number, n: number): number {
  const dp = Array.from({ length: m }, () =>
    Array.from({ length: n }, () => 0),
  );

  for (let i = 0; i < m; i++) {
    dp[i][0] = 1;

    for (let j = 1; j < n; j++) {
      dp[i][j] = dp[i][j - 1] + (i > 0 ? dp[i - 1][j] : 0);
    }
  }

  return dp[m - 1][n - 1];
}
// @leet end
